model/repository/MonitoringRepository.php

```diff
@@ -460,23 +460,41 @@ protected function prepareOrderStmt($order): string
     {
         $order = explode(',', $order);
         $result = [];
-        $primaryTableColumns = $this->getPrimaryColumns();
-        foreach ($order as $ruleNum => $orderRule) {
-            preg_match('/([a-zA-Z_][a-zA-Z0-9_]*)\s?(asc|desc)?\s?(string|numeric)?/i', $orderRule, $ruleParts);
+        foreach ($order as $orderRule) {
+            $result[] = $this->buildSingleOrderRule($orderRule);
+        }
 
-            if (!in_array($ruleParts[1], $primaryTableColumns)) {
-                $colName = $ruleParts[1];
-                $this->queryParams[] = $colName;
+        return implode(', ', $result);
+    }
+
+    private function buildSingleOrderRule(string $orderRule): string
+    {
+        preg_match(
+            '/([a-zA-Z_][a-zA-Z0-9_]*)\s?(asc|desc)?\s?(string|numeric)?/i',
+            $orderRule,
+            $ruleParts
+        );
+
+        if (!in_array($ruleParts[1], $this->getPrimaryColumns(), true)) {
+            $colName = $ruleParts[1];
+            if (in_array($this->getPlatformName(), ['mysql', 'sqlite'])) {
+                $colName = sprintf('JSON_EXTRACT(t.%s, \'$.%s\')', self::COLUMN_EXTRA_DATA, $colName);
             } else {
-                $sortingColumn = $ruleParts[1];
+                $colName = sprintf('t.%s ->> \'%s\'', self::COLUMN_EXTRA_DATA, $colName);
             }
-
-            $result[] = isset($ruleParts[3]) && $ruleParts[3] === 'numeric'
-                ? sprintf("cast(nullif(%s, '') as decimal) %s", $sortingColumn, $ruleParts[2])
-                : sprintf('%s %s', $sortingColumn, isset($ruleParts[2]) ? $ruleParts[2] : 'ASC');
+            $sortingColumn = $colName;
+        } else {
+            $sortingColumn = $ruleParts[1];
         }
 
-        return implode(', ', $result);
+        // Hotfix for sorting by authorised_by column which can contain a curd number or user URI
+        $numericPart = in_array($this->getPlatformName(), ['mysql', 'sqlite'])
+            ? sprintf("cast(NULLIF(%s, '') as SIGNED) %s", $sortingColumn, $ruleParts[2])
+            : sprintf("cast(NULLIF(regexp_replace(%s, '\D', '', 'g'), '') as decimal) %s", $sortingColumn, $ruleParts[2]);
+
+        return isset($ruleParts[3]) && $ruleParts[3] === 'numeric'
```

Comment:
I'd probably add the `in_array($ruleParts[1], $this->getPrimaryColumns(), true)` condition here too.

There's no way to cast the json column contents I guess, or is there?

Reply:
it's already cast to text on https://github.com/oat-sa/extension-tao-proctoring/pull/920/files#diff-f8104ec72a06b9191a4d5ed139f5addc39c1d463ad5ad808d27aa5c1c0c44e76R483 and that allow to cast it to decimal